src/passes/CodeFolding.cpp

```diff
@@ -243,6 +243,10 @@ struct CodeFolding : public WalkerPass<ControlFlowWalker<CodeFolding>> {
     if (!curr->ifFalse) {
       return;
     }
+    if (curr->ifTrue->type.isConcrete()) {
+      // We don't support folding tails that produce values.
+      return;
+    }
```

Comment:
We don't support that for brs, but for ifs I think we can? New line 259 has the logic to keep the concrete type the same.

Reply:
The comment on lines 502 to 508 explains one of the places that assumes we never have to handle folding concrete types.

The comment `// we must ensure we present the same type as the if had` on line 258 is also wrong; if an `if` is unreachable because of an unreachable condition, the `ifTrue` arm might have some other type.

Reply:
The comment on line 502 looks correct to me, but it relates to blocks and brs, and not ifs, unless I'm confused?

Good point about unreachability, it is possible we aren't handling that well enough here. But I think we can add that without removing this optimization.

But with that said, I see we have that merging of arms logic in OptimizeInstructions:

https://github.com/WebAssembly/binaryen/blob/485ac7da1e72d1b6d4305317eb7fe1a9dd046c0a/src/passes/OptimizeInstructions.cpp#L1140

So this looks like a redundant optimization actually, and perhaps we can remove it from here. But it is possible we don't handle unreachability fully there, so that might need fixing.

Reply:
The merging of if arms is handled by the same code that optimizes blocks and brs, so it’s not really possible that we support concrete types for one and not the others. It just happened to work out by accident until now because of the extra refinalization.

I’ll look into whether we can remove the optimizations of Ifs here completely or support concrete types for all folded branches.

Reply:
I don't think the folding in OptimizeInstructions can completely replace the optimizations here because it only works if the arms are completely identical, whereas this pass only requires identical suffixes.

Reply:
Oh, it looks like line 251 compares the entire ifTrue and ifFalse arms, while 263+ looks at suffixes. Was your change needed for one of the two?